src/main/java/seedu/pill/util/ItemMap.java

```diff
@@ -0,0 +1,335 @@
+package seedu.pill.util;
+
+import seedu.pill.exceptions.ExceptionMessages;
+import seedu.pill.exceptions.PillException;
+
+import java.time.LocalDate;
+import java.util.Map;
+import java.util.TreeSet;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.List;
+import java.util.LinkedHashMap;
+import java.util.logging.Logger;
+import java.util.stream.IntStream;
+
+/**
+ * Represents a list of items and provides methods to add, delete, list, and edit items.
+ */
+public class ItemMap implements Iterable<Map.Entry<String, TreeSet<Item>>> {
+    private static final Logger LOGGER = PillLogger.getLogger();
+    Map<String, TreeSet<Item>> items;
+
+    /**
+     * Constructor for ItemMap.
+     * Initializes the internal Map to store items.
+     */
+    public ItemMap() {
+        this.items = new LinkedHashMap<>();
+        LOGGER.info("New ItemMap instance created");
+    }
+
+    public boolean isEmpty() {
+        return this.items.isEmpty();
+    }
+
+    @Override
+    public Iterator<Map.Entry<String, TreeSet<Item>>> iterator() {
+        return items.entrySet().iterator();
+    }
+
+    /**
+     * Adds a new item to the list.
+     *
+     * @param newItem The item to be added.
+     */
+    public void addItem(Item newItem) {
```

Comment:
Maybe can look into splitting this method into several smaller methods as it is currently quite long